### ISSUE TITLE
Failed Robolectric 4 unit tests running with Gradle

### ISSUE THREAD

**Issue by nikoskyr** [robolectric/robolectric#4026]
Hello,
### Description
I faced a problem when I tried to run the Robolectric unit tests with gradle. The error I got is the following, not descriptive log:

```
MyActivityTest > shouldNotBeNull FAILED
    java.lang.NullPointerException
MyActivityTest: sdk=28; resources=binary
java.lang.RuntimeException
        Caused by: android.content.pm.PackageParser$PackageParserException
            Caused by: java.lang.NullPointerException
```

### Steps to Reproduce
```groovy
testOptions {
        unitTests {
            //returnDefaultValues = true
            includeAndroidResources = true
        }
    }
```
Add the following line to `gradle.properties`:
```properties
android.enableUnitTestBinaryResources=true
```

```java
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import androidx.test.ext.junit.runners.AndroidJUnit4;
import androidx.test.rule.ActivityTestRule;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.isEnabled;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static org.junit.Assert.assertNotNull;

@RunWith(AndroidJUnit4.class)
public class MyActivityTest {


    private MyActivity activity;

    @Rule
    public ActivityTestRule< MyActivity > mActivityRule = new ActivityTestRule<>(MyActivity.class);

    @Before
    public void setUp() {
        activity = mActivityRule.getActivity();
    }

    @Test
    public void shouldNotBeNull() throws Exception {
        assertNotNull( activity );
    }
```
Run the tests with:
`./gradlew testProductionDebugUnitTest`

### Robolectric & Android Version
Robolectric 4 beta 1 - Android 3.2.1 and Gradle plugin 3.2.1

**Comment by xian** [comment]
@nikoskyr could you include a full stacktrace? There should be one in the ftp test results, which are usually written as an html report in your build directory.

**Comment by DevHugo** [comment]
I got the same type of error : 

```
java.lang.RuntimeException: android.content.pm.PackageParser$PackageParserException: Failed to read manifest from /Users/hugo/dev/java/beschool-android/app/build/intermediates/apk_for_local_test/autoTestDebugUnitTest/packageAutoTestDebugUnitTestForUnitTest/apk-for-local-test.ap_

	at org.robolectric.shadows.ShadowPackageParser.callParsePackage(ShadowPackageParser.java:57)
	at org.robolectric.android.internal.ParallelUniverse.setUpApplicationState(ParallelUniverse.java:142)
	at org.robolectric.RobolectricTestRunner.beforeTest(RobolectricTestRunner.java:372)
	at org.robolectric.internal.SandboxTestRunner$2.evaluate(SandboxTestRunner.java:252)
	at org.robolectric.internal.SandboxTestRunner.runChild(SandboxTestRunner.java:130)
	at org.robolectric.internal.SandboxTestRunner.runChild(SandboxTestRunner.java:42)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.robolectric.internal.SandboxTestRunner$1.evaluate(SandboxTestRunner.java:84)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.application.AppMainV2.main(AppMainV2.java:131)
Caused by: android.content.pm.PackageParser$PackageParserException: Failed to read manifest from /Users/hugo/dev/java/beschool-android/app/build/intermediates/apk_for_local_test/autoTestDebugUnitTest/packageAutoTestDebugUnitTestForUnitTest/apk-for-local-test.ap_
	at android.content.pm.PackageParser.parseBaseApk(PackageParser.java:1353)
	at android.content.pm.PackageParser.parseMonolithicPackage(PackageParser.java:1299)
	at android.content.pm.PackageParser.parsePackage(PackageParser.java:1022)
	at android.content.pm.PackageParser.parsePackage(PackageParser.java:1042)
	at org.robolectric.shadows.ShadowPackageParser.callParsePackage(ShadowPackageParser.java:31)
	at org.robolectric.android.internal.ParallelUniverse.setUpApplicationState(ParallelUniverse.java:142)
	at org.robolectric.RobolectricTestRunner.beforeTest(RobolectricTestRunner.java:372)
	at org.robolectric.internal.SandboxTestRunner$2.evaluate(SandboxTestRunner.java:252)
	at org.robolectric.internal.SandboxTestRunner.runChild(SandboxTestRunner.java:130)
	at org.robolectric.internal.SandboxTestRunner.runChild(SandboxTestRunner.java:42)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.robolectric.internal.SandboxTestRunner$1.evaluate(SandboxTestRunner.java:84)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
	... 1 more
Caused by: java.lang.NullPointerException
	at org.robolectric.res.android.CppAssetManager2.FindEntry(CppAssetManager2.java:678)
	at org.robolectric.res.android.CppAssetManager2.GetResource(CppAssetManager2.java:841)
	at org.robolectric.res.android.CppAssetManager2.ResolveReference(CppAssetManager2.java:899)
	at org.robolectric.res.android.AttributeResolution9.RetrieveAttributes(AttributeResolution9.java:484)
	at org.robolectric.shadows.ShadowArscAssetManager9.nativeRetrieveAttributes(ShadowArscAssetManager9.java:1454)
	at android.content.res.AssetManager.nativeRetrieveAttributes(AssetManager.java)
	at android.content.res.AssetManager.retrieveAttributes(AssetManager.java:1012)
	at android.content.res.Resources.obtainAttributes(Resources.java:1814)
	at android.content.res.Resources$GeneratedProxy/361380654.obtainAttributes(Unknown Source)
	at org.robolectric.shadows.ShadowResources.obtainAttributes(ShadowResources.java:90)
	at android.content.res.Resources.obtainAttributes(Resources.java)
	at android.content.pm.PackageParser.parseMetaData(PackageParser.java:5318)
	at android.content.pm.PackageParser.parseProviderTags(PackageParser.java:4961)
	at android.content.pm.PackageParser.parseProvider(PackageParser.java:4927)
	at android.content.pm.PackageParser.parseBaseApplication(PackageParser.java:3671)
	at android.content.pm.PackageParser.parseBaseApkCommon(PackageParser.java:2053)
	at android.content.pm.PackageParser.parseBaseApk(PackageParser.java:1942)
	at android.content.pm.PackageParser.parseBaseApk(PackageParser.java:1337)
	... 22 more
```

**Comment by PaulWoitaschek** [comment]
Exact same error message as posted above :+1:

**Comment by nikoskyr** [comment]
Hello @xian,

The stack trace I got is identical to the one DevHugo has previously shared. 
When I deleted the `android.enableUnitTestBinaryResources=true` the cases failed again, but this time with an error log produced be leak Canary. 

However, this was strange because I have added the following line in my build.gradle
`testImplementation "com.squareup.leakcanary:leakcanary-android-no-op:$rootProject.leakCanaryVersion"`

The problem was only solved, when I commented the entire line that initializes leak canary.

Please let me know if you need some extra logs.

Best regards,
Nikos

**Comment by nikoskyr** [comment]
Hello @xian,

I have the same stacktrace that DevHugo has already copied.

However, when I deleted the following line, unit tests failed with another reason. 
`android.enableUnitTestBinaryResources=true`

I searched a little bit and that reason was leak canary, but it was strange because I have the following configuration in my build.gradle.
`testImplementation "org.json:json:$rootProject.testJsonVersion"
    testImplementation "com.squareup.okhttp3:mockwebserver:$rootProject.okHttpVersion"
    testImplementation "org.mockito:mockito-core:$rootProject.mockitoCoreVersion"
    testImplementation "org.hamcrest:hamcrest-all:$rootProject.hamcrestVersion"
    testImplementation "org.powermock:powermock-api-mockito2:$rootProject.powerMockVersion"
    testImplementation "org.powermock:powermock-module-junit4-rule-agent:$rootProject.powerMockVersion"
    testImplementation "org.powermock:powermock-module-junit4-rule:$rootProject.powerMockVersion"
    testImplementation "org.powermock:powermock-module-junit4:$rootProject.powerMockVersion"
    testImplementation("androidx.test.ext:junit:$rootProject.androidXJunitVersion") {
        exclude group: "org.junit"
    }
    testImplementation "androidx.test:runner:$rootProject.runnerVersion"
    testImplementation "androidx.test:rules:$rootProject.rulesVersion"
    testImplementation "androidx.test.espresso:espresso-core:$rootProject.espressoVersion"
    testImplementation "org.robolectric:robolectric:$rootProject.robolectricVersion"
    testImplementation "com.squareup.leakcanary:leakcanary-android-no-op:$rootProject.leakCanaryVersion"`

Eventually, unit tests passed when I commented the line in my code that initializes leak canary.

If you need extra logs please let me know.

Best Regards,
Nikos

**Comment by xian** [comment]
Hey, thanks for the diagnostic hints, but I'm still having trouble reproducing... could you attach a full (but minimal) manifest that exhibits the problem?

**Comment by nikoskyr** [comment]
Hello @xian,

Please find attached the manifest. I deleted some stuff for obvious reasons.
This is the debug manifest
`<?xml version="1.0" encoding="utf-8"?>
<manifest
    xmlns:android="http://schemas.android.com/apk/res/android"
    android:versionCode="9"
    android:versionName="5"
    android:compileSdkVersion="28"
    android:compileSdkVersionCodename="9"
    package="xxx"
    platformBuildVersionCode="9"
    platformBuildVersionName="5">

    <uses-sdk
        android:minSdkVersion="23"
        android:targetSdkVersion="28" />

    <uses-feature
        android:name="android.hardware.camera.front"
        android:required="false" />

    <uses-feature
        android:name="android.hardware.camera.autofocus"
        android:required="false" />

    <uses-feature
        android:name="android.hardware.camera.flash"
        android:required="false" />

    <uses-feature
        android:name="android.hardware.screen.landscape"
        android:required="false" />

    <uses-feature
        android:name="android.hardware.wifi"
        android:required="false" />

    <application
        android:theme="@ref/xxxx"
        android:label="@ref/xxxx"
        android:icon="@ref/xxxx"
        android:name="MainApplication"
        android:debuggable="true"
        android:allowBackup="false"
        android:supportsRtl="true"
        android:roundIcon="@ref/xxxx"
        android:appComponentFactory="androidx.core.app.CoreComponentFactory">

        <activity
            android:label="@string/xxxx"
            android:name="MyActivity"
            android:exported="true"
            android:screenOrientation="1" />

            <intent-filter>

                <action
                    android:name="android.intent.action.MAIN" />

                <category
                    android:name="android.intent.category.LAUNCHER" />
            </intent-filter>
        </activity>

        <service
            android:name="com.squareup.leakcanary.internal.HeapAnalyzerService"
            android:enabled="false"
            android:process=":leakcanary" />

        <service
            android:name="com.squareup.leakcanary.DisplayLeakService"
            android:enabled="false"
            android:process=":leakcanary" />

        <provider
            android:name="com.squareup.leakcanary.internal.LeakCanaryFileProvider"
            android:exported="false"
            android:authorities="com.squareup.leakcanary.fileprovider.xxx"
            android:grantUriPermissions="true">

            <meta-data
                android:name="android.support.FILE_PROVIDER_PATHS"
                android:resource="@ref/0x7f110000" />
        </provider>

        <activity
            android:theme="@ref/0x7f0f0204"
            android:label="@ref/0x7f0e00fb"
            android:icon="@ref/0x7f0c0002"
            android:name="com.squareup.leakcanary.internal.DisplayLeakActivity"
            android:enabled="false"
            android:process=":leakcanary"
            android:taskAffinity="com.squareup.leakcanary.xxx">

            <intent-filter>

                <action
                    android:name="android.intent.action.MAIN" />

                <category
                    android:name="android.intent.category.LAUNCHER" />
            </intent-filter>
        </activity>

        <activity
            android:theme="@ref/0x7f0f0205"
            android:label="@ref/0x7f0e0114"
            android:icon="@ref/0x7f0c0002"
            android:name="com.squareup.leakcanary.internal.RequestStoragePermissionActivity"
            android:enabled="false"
            android:process=":leakcanary"
            android:taskAffinity="xxx"
            android:excludeFromRecents="true"
            android:roundIcon="@ref/0x7f0c0002" />
    </application>
</manifest>
`

**Comment by xian** [comment]
@nikoskyr, sorry but I'm still unable to reproduce the failure. I suspect this might be related to resource qualifiers; do you have multiple `provider_paths.xml` resources perhaps?

A copy of your app's `resources.ap_` would be super helpful if you're able to share... you could email me at christianw at g...le.com if you prefer.

**Comment by DevHugo** [comment]
I discuss with my team, if we can opensource the project. I let you know as soon as possible.

**Comment by calvarez-ov** [comment]
Here's a sample app: https://github.com/libon/RobolectricManifestIssue

The bug requires 3 conditions:
* Have this in the manifest, on the `<application>` tag:

```xml
android:networkSecurityConfig="@xml/network_security_config"
```

* In `gradle.properties`, enable binary resources:
```properties
android.enableUnitTestBinaryResources=true
```

* In `app/build.gradle`, add this in the `android` section:
```gradle
    testOptions {
        unitTests {
            includeAndroidResources = true
        }
    }
```

If any of these conditions is removed, tests pass.

**Comment by calvarez-ov** [comment]
Indeed, the networkSecurityConfig is just one example of a manifest config that can trigger this. Another example I saw in our project was a `FileProvider`.

**Comment by tomhermann** [comment]
I ran into the same issue today on my project. In my case, having Leak Canary no-op on the classpath triggered the issue.

**Comment by xian** [comment]
Fix coming with 4.0.1, thanks for the help diagnosing!

**Comment by mikehardy** [comment]
Thanks for the fix @xian - I can confirm 4.0.1 works with binary resource loading in AnkiDroid. Strangely, in order to use binary resource loading I had to add an override-library entry like so in an AndroidManifest.xml in the test directory:

```
<?xml version="1.0" encoding="utf-8"?>
<manifest
    xmlns:tools="http://schemas.android.com/tools"
    package="com.ichi2.anki.api">
    <uses-sdk tools:overrideLibrary="androidx.test"/>
</manifest>
```

because we support an SDK below 14 - but when I don't use binary resource loading, it wasn't necessary. So it seems there is still some difference between what and how things get built with binary resources vs legacy style, but this test-only override is acceptable, so we're working now without deprecation on robolectric 4.x

**Comment by xian** [comment]
@mikehardy — glad to hear it's working!

Yep, legacy and binary modes are vastly different; in binary mode you'll see warnings and errors from the Android build toolchain in addition to Robolectric, but everything should be much more faithful to proper Android behavior. My guess is that this error is coming from `aapt`, before Robolectric is even invoked. If you can give me a simple test project that shows the problem I can dig in further.

Thanks!
